controller/object_store_controller.go

```diff
@@ -0,0 +1,1250 @@
+		logrus.Errorf("Object Store running but PVC not bound")
+		return err
+	}
+
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopping(store *longhorn.ObjectStore) (err error) {
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err != nil {
+		return errors.Wrap(err, "failed find Deployment to stop")
+	} else if (*dpl).Spec.Replicas != nil && *((*dpl).Spec.Replicas) != 0 {
+		(*dpl).Spec.Replicas = util.Int32Ptr(0)
+		_, err = osc.ds.UpdateDeployment(dpl)
+		return err
+	} else if dpl.Status.AvailableReplicas > 0 {
+		return nil // wait for shutdown
+	}
+
+	logrus.Infof("object store %v is now stopped", store.Name)
+	store.Status.State = longhorn.ObjectStoreStateStopped
+	return nil
+}
+
+func (osc *ObjectStoreController) handleStopped(store *longhorn.ObjectStore) (err error) {
+	if store.Spec.TargetState == longhorn.ObjectStoreStateRunning {
+		logrus.Infof("object store %v is now starting", store.Name)
+		store.Status.State = longhorn.ObjectStoreStateStarting
+		return nil
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) handleTerminating(store *longhorn.ObjectStore) (err error) {
+	// The resources are created in the order:
+	// Volume -> PV -> PVC -> Deployment -> Service
+	// so we tear them down in reverse:
+	// Service -> Deployment -> PVC -> PV -> Volume
+	// Once that is done we can remove the finalizer, which allows the K8s API to
+	// remove the object store resource as well.
+	_, err = osc.ds.GetService(osc.namespace, store.Name)
+	if err == nil {
+		return osc.ds.DeleteService(osc.namespace, store.Name)
+	} else if !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	_, err = osc.ds.GetDeployment(store.Name)
+	if err == nil {
+		return osc.ds.DeleteDeployment(store.Name)
+	} else if !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	_, err = osc.ds.GetPersistentVolumeClaim(osc.namespace, store.Name)
+	if err == nil {
+		return osc.ds.DeletePersistentVolumeClaim(osc.namespace, store.Name)
+	} else if !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	_, err = osc.ds.GetPersistentVolume(store.Name)
+	if err == nil {
+		osc.ds.DeletePersistentVolume(store.Name)
+	} else if !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	_, err = osc.ds.GetVolume(store.Name)
+	if err == nil {
+		osc.ds.DeleteVolume(store.Name)
+	} else if !datastore.ErrorIsNotFound(err) {
+		return err
+	}
+
+	// cleanup all secrets with matching labels.
+	labels := types.GetBaseLabelsForSystemManagedComponent()
+	labels[types.GetLonghornLabelComponentKey()] = types.LonghornLabelObjectStore
+	labels[types.GetLonghornLabelKey(types.LonghornLabelObjectStore)] = store.Name
+
+	secrets, err := osc.ds.ListSecretsByLabels(osc.namespace, labels)
+	if err != nil {
+		return err
+	}
+
+	for _, secret := range secrets {
+		osc.ds.DeleteSecret(osc.namespace, secret.Name)
+	}
+
+	if len(store.ObjectMeta.Finalizers) != 0 {
+		return osc.ds.RemoveFinalizerForObjectStore(store)
+	}
+
+	return nil
+}
```

Comment:
suggesting to directly delete the object and check the returned error instead of getting the object then delete
```suggestion
func (osc *ObjectStoreController) handleTerminating(store *longhorn.ObjectStore) (err error) {
	// The resources are created in the order:
	// Volume -> PV -> PVC -> Deployment -> Service
	// so we tear them down in reverse:
	// Service -> Deployment -> PVC -> PV -> Volume
	// Once that is done we can remove the finalizer, which allows the K8s API to
	// remove the object store resource as well.

	if err := osc.ds.DeleteService(osc.namespace, store.Name); err != nil && !datastore.ErrorIsNotFound(err) {
		return err
	}
	if err := osc.ds.DeleteDeployment(store.Name); err != nil && !datastore.ErrorIsNotFound(err) {
		return err
	}
	if err := osc.ds.DeletePersistentVolumeClaim(osc.namespace, store.Name); err != nil && !datastore.ErrorIsNotFound(err) {
		return err
	}
	if err := osc.ds.DeletePersistentVolume(store.Name); err != nil && !datastore.ErrorIsNotFound(err) {
		return err
	}
	if err := osc.ds.DeleteVolume(store.Name); err != nil && !datastore.ErrorIsNotFound(err) {
		return err
	}

	// cleanup all secrets with matching labels.
	labels := types.GetBaseLabelsForSystemManagedComponent()
	labels[types.GetLonghornLabelComponentKey()] = types.LonghornLabelObjectStore
	labels[types.GetLonghornLabelKey(types.LonghornLabelObjectStore)] = store.Name
	secrets, err := osc.ds.ListSecretsByLabels(osc.namespace, labels)
	if err != nil {
		return err
	}
	for _, secret := range secrets {
		if err := osc.ds.DeleteSecret(osc.namespace, secret.Name); err != nil && !datastore.ErrorIsNotFound(err) {
			return err
		}
	}

	return osc.ds.RemoveFinalizerForObjectStore(store)
}
```